duckduckgo/p5-app-duckpan
# Add Fathead support for DuckPAN Query & Server

/cc @zachthompson @jbarrett @jdorweiler 

These changes add support to `duckpan query` and `duckpan server` for Fathead features used by open source Fatheads.

You can query the exact names of articles in the Output.txt file to see the result that will display. If an article is a redirect, it will follow the redirect and show the correct result. Disambiguations and Abstract Images are also supported.

Category Pages and External Links (infobox) support has not been implemented yet. But I plan to implement the latter soon. Also Fathead Triggers are not yet supported but I will add that soon as well, using the IA Page triggers. However, internally I don't believe we consider the triggers defined on the IA page (though we should!)

This approach ignores the DDG infrastructure and request handlers, which Spice and Goodies use. That system is heavily dependant on loading Fathead Perl modules into Blocks, but we removed the Perl modules long ago as they were only there to contain Fathead Metadata.

One other missing feature is the internal processing of the output.txt -- Internally we process each line of the output.txt and generate several variations of the Article title, e.g. for "background-color" we also create "background color" and "backgroundcolor".

I noticed that some Fathead parsing scripts implement this functionality themselves. We should probably make it known to developers that we'll do this work for them to simplify the parsers a bit.

Any feedback is welcome. I've not used DBI before so I wasn't sure about the best way to get the output.txt into a DB for searching. I've used **DBD::CSV** which interprets the output.txt (tab delimited) as a CSV and pulls it in to memory for SQL querying. I'm not sure how well this will work for very large output.txt files. We can easily test this though.

### How to Test

1. `git checkout` this branch
2. Clone the Fathead repository: `git clone git@github.com:duckduckgo/zeroclickinfo-fathead.git`
3. Start the server: `$ duckpan -I ../p5-app-duckpan/lib/ server <instant_answer_id>`
    e.g. `$ duckpan -I ../p5-app-duckpan/lib/ server mdn_css`
4. Enter one of the article titles from the associated Output.txt as a search query to see the content displayed.
  - e.g. "auth_basic" from https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/lib/fathead/nginx/output.txt

-------

To view the changes made while ignoring whitespace changes: https://github.com/duckduckgo/p5-app-duckpan/pull/358/files?w=0
